File: src/main/java/icu/samnyan/aqua/sega/chusan/dao/userdata/UserDataRepository.java
package icu.samnyan.aqua.sega.chusan.dao.userdata;

import icu.samnyan.aqua.net.utils.GenericUserDataRepo;
import icu.samnyan.aqua.sega.chusan.model.userdata.UserData;
import icu.samnyan.aqua.sega.general.model.Card;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.Optional;

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Repository("ChusanUserDataRepository")
public interface UserDataRepository extends JpaRepository<UserData, Long>, GenericUserDataRepo {

    UserData findByCard(Card card);

    Optional<UserData> findByCard_ExtId(Long extId);

    @Override
    @Query("select count(*) from Maimai2UserData where playerRating > :rating")
    long getRanking(int rating);
}

File: src/main/java/icu/samnyan/aqua/sega/maimai2/dao/userdata/UserDataRepository.kt
package icu.samnyan.aqua.sega.maimai2.dao.userdata

import icu.samnyan.aqua.net.games.GenericUserDataRepo
import icu.samnyan.aqua.sega.general.model.Card
import icu.samnyan.aqua.sega.maimai2.model.userdata.UserDetail
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import java.util.*


@Repository("Maimai2UserDataRepository")
interface UserDataRepository : GenericUserDataRepo<UserDetail, Long> {
    override fun findByCard(card: Card): UserDetail?

    fun findByCardExtId(userId: Long): Optional<UserDetail>

    @Transactional
    fun deleteByCard(card: Card)

    @Query("select count(*) from Maimai2UserData where playerRating > :rating")
    override fun getRanking(rating: Int): Long
}
